source/geod24/concurrency.d

```diff
@@ -934,76 +1007,616 @@ protected:
         {
-            assert(info_fiber !is null, "This Fiber does not belong to FiberScheduler");
-            assert(info_fiber.sem is null, "This Fiber already has a registered FiberBinarySemaphore");
-            info_fiber.sem = this;
+            if (cas(&this.consumed, false, true))
+            {
+                this.id = id;
+                this.success = success_in;
+                return true;
+            }
+            return false;
+        }
+
+        /***********************************************************************
+
+            Returns if SelectState is already consumed or not
+
+        ***********************************************************************/
 
+        bool isConsumed() nothrow
+        {
+            return atomicLoad(this.consumed);
         }
 
-        void wait() nothrow
+        /// ID of the select call that consumed this SelectState
+        int id;
+
+        /// Success/failure state of the select call with id of `id`
+        bool success;
+
+    private:
+        /// Indicates if this SelectState is consumed or not
+        shared(bool) consumed;
+    }
+
+    /***********************************************************************
+
+        A golang style channel implementation with buffered and unbuffered
+        operation modes
+
+        Intended to be used between Fibers
+
+    ***********************************************************************/
+
+    final class Channel (T)
+    {
+
+        /***********************************************************************
+
+            Constructs a Channel
+
+            Param:
+                max_size = Maximum amount of T a Channel can buffer
+                                0 -> Unbuffered operation
+                                Positive integer -> Buffered operation
+
+        ***********************************************************************/
+
+        this(ulong max_size = 0) nothrow
         {
-            this.registerToInfoFiber();
-            FiberScheduler.yield();
-            scope (exit) this.notified = false;
+            this.max_size = max_size;
+            this.lock = new Mutex;
         }
 
-        bool wait(Duration period) nothrow
+        /***********************************************************************
+
+            Write a message to the Channel
+
+            Unbuffered mode:
+
+                If a reader is already blocked on the Channel, writer copies the
+                message to reader's buffer and wakes up the reader by yielding
+
+                If no reader is ready in the wait queue, writer appends itself
+                to write wait queue and blocks
+
+            Buffered mode:
+
+                If a reader is already blocked on the Channel, writer copies the
+                message to reader's buffer and wakes up the reader by yielding.
+                After execution comes back to the writer, it returns immediatly.
```

Comment:
Technically this differs from the Golang semantic. In Go, writing to an unbuffered channel *always* block the caller (either really blocking, or hand off to pending reader), but writing to a buffered channel never blocks for the writer if the channel isn't full.

Example code:
```Go
package main

import (
	"fmt"
)

func receiver(ch chan string) {
	foo := <- ch
	fmt.Println("Received: ", foo)
}

func main() {
         // Create an unbuffered channel
	ch := make(chan string)
	// Create a buffered channel
        ch := make(chan string, 2)
	go receiver(ch)
	ch <- "World"
	ch <- "World"	 // Remove me if using unbuffered channel
	fmt.Println("Hello, playground")
}
```

If you run it on the playground, [you can see](https://play.golang.org/p/ZqcJvc1pCsk) that it only prints "Hello playground", never prints the "Hello" sent.

Now, the title of the issue mentioned "Go style concurrency" not because we need to emulate Go in every aspect, but because we wanted to provide a directional, pipe based API. So while we don't have to emulate every aspects of it, I wonder if this  semantic is actually relied on somewhere, and how useful it actually is.

Thinking about it, it would make sense to have a buffered channel when your writer has high throughput, and unbuffered channel when the reader does (because of the hands off semantic).

Reply:
This is solely because the `FiberBinarySemaphore` yields the running Fiber on `notify()` calls. I tried to change that before, but the agora tests failed.

Reply:
This is closer to Golang behavior with the new FiberScheduler in this [branch](https://github.com/omerfirmak/localrest/tree/go-style-sched).